File: src/org/elixir_lang/templates/ElixirContextType.java
package org.elixir_lang.templates;

import com.intellij.codeInsight.template.EverywhereContextType;
import com.intellij.codeInsight.template.TemplateContextType;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiWhiteSpace;
import com.intellij.psi.util.PsiUtilCore;
import org.elixir_lang.ElixirLanguage;
import org.jetbrains.annotations.NotNull;

public class ElixirContextType extends TemplateContextType {

    ElixirContextType() {
        super("ELIXIR_CODE", "Elixir", EverywhereContextType.class);
    }

    @Override
    public boolean isInContext(@NotNull PsiFile psiFile, int offset) {
        if (!PsiUtilCore.getLanguageAtOffset(psiFile, offset).isKindOf(ElixirLanguage.INSTANCE)) return false;
        PsiElement element = psiFile.findElementAt(offset);
        return !(element instanceof PsiWhiteSpace) && element != null;
    }
}

File: src/org/elixir_lang/templates/ElixirContextType.kt
package org.elixir_lang.templates

import com.intellij.codeInsight.template.TemplateActionContext
import com.intellij.codeInsight.template.TemplateContextType
import com.intellij.psi.PsiWhiteSpace
import com.intellij.psi.util.PsiUtilCore
import org.elixir_lang.ElixirLanguage

class ElixirContextType internal constructor() :
    TemplateContextType("Elixir") {

    override fun isInContext(templateActionContext: TemplateActionContext): Boolean {
        val psiFile = templateActionContext.file
        val offset = templateActionContext.startOffset

        return if (!PsiUtilCore.getLanguageAtOffset(psiFile, offset).isKindOf(ElixirLanguage)) {
            false
        } else {
            val element = psiFile.findElementAt(offset)
            element !is PsiWhiteSpace && element != null
        }
    }
}
